File: backends/fastapi/app/schemas/capsule.py
"""胶囊相关 DTO。"""

from __future__ import annotations

import uuid
from datetime import datetime, timedelta, timezone

from pydantic import BaseModel, Field, field_validator

from app.schemas.user import UserBrief


class CapsuleBase(BaseModel):
    id: uuid.UUID
    code: str
    title: str
    creator: UserBrief
    openAt: datetime
    createdAt: datetime
    inPlaza: bool
    favoriteCount: int
    isOpened: bool


class CapsuleListItem(CapsuleBase):
    favoritedByMe: bool = False
    favoritedAt: datetime | None = None


class CapsuleDetail(CapsuleBase):
    content: str | None = None
    favoritedByMe: bool = False


class CreateCapsuleRequest(BaseModel):
    title: str = Field(min_length=1, max_length=60)
    content: str = Field(min_length=1, max_length=5000)
    openAt: datetime
    inPlaza: bool = True

    @field_validator("openAt")
    @classmethod
    def _open_at_bounds(cls, v: datetime) -> datetime:
        now = datetime.now(timezone.utc)
        # 归一：naive → 视为 UTC
        if v.tzinfo is None:
            v = v.replace(tzinfo=timezone.utc)
        if v < now + timedelta(seconds=60):
            raise ValueError("openAt 必须晚于当前时间 60 秒以上")
        if v > now + timedelta(days=365 * 10):
            raise ValueError("openAt 不得超出当前时间 10 年")
        return v


class FavoriteRequest(BaseModel):
    capsuleId: uuid.UUID


class FavoriteResult(BaseModel):
    capsuleId: uuid.UUID
    favoriteCount: int
    favoritedAt: datetime

File: backends/gin/internal/dto/capsule.go
package dto

import "time"

// CapsuleBase 胶囊公共字段。
type CapsuleBase struct {
	ID            string    `json:"id"`
	Code          string    `json:"code"`
	Title         string    `json:"title"`
	Creator       UserBrief `json:"creator"`
	OpenAt        time.Time `json:"openAt"`
	CreatedAt     time.Time `json:"createdAt"`
	InPlaza       bool      `json:"inPlaza"`
	FavoriteCount int       `json:"favoriteCount"`
	IsOpened      bool      `json:"isOpened"`
}

// CapsuleDetail 单个胶囊详情（含内容）。
type CapsuleDetail struct {
	CapsuleBase
	Content       *string `json:"content"`
	FavoritedByMe bool    `json:"favoritedByMe"`
}

// CapsuleListItem 列表项（含收藏状态与内容预览）。
type CapsuleListItem struct {
	CapsuleBase
	FavoritedByMe  bool       `json:"favoritedByMe"`
	FavoritedAt    *time.Time `json:"favoritedAt,omitempty"`
	ContentPreview *string    `json:"contentPreview"`
}

// CreateCapsuleRequest 创建胶囊请求体。
type CreateCapsuleRequest struct {
	Title   string    `json:"title"   binding:"required,min=1,max=60"`
	Content string    `json:"content" binding:"required,min=1,max=5000"`
	OpenAt  time.Time `json:"openAt"  binding:"required"`
	InPlaza *bool     `json:"inPlaza"`
}

// FavoriteRequest 收藏请求体。
type FavoriteRequest struct {
	CapsuleID string `json:"capsuleId" binding:"required"`
}

// FavoriteResult 收藏操作结果。
type FavoriteResult struct {
	CapsuleID     string    `json:"capsuleId"`
	FavoriteCount int       `json:"favoriteCount"`
	FavoritedAt   time.Time `json:"favoritedAt"`
}

// Avatar 头像目录项。
type Avatar struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	PrimaryColor string `json:"primaryColor"`
	SvgURL       string `json:"svgUrl"`
}
